source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc

```diff
@@ -385,164 +357,109 @@ bool Filter::parseTlvs(const std::vector<uint8_t>& tlvs) {
     }
 
     idx += tlv_value_length;
-    ASSERT(idx <= tlvs.size());
+    ASSERT(idx <= len);
   }
   return true;
 }
 
-ReadOrParseState Filter::readExtensions(Network::IoHandle& io_handle) {
-  // Parse and discard the extensions if this is a local command or there's no TLV needs to be saved
-  // to metadata.
-  if (proxy_protocol_header_.value().local_command_ || 0 == config_->numberOfNeededTlvTypes()) {
-    // buf_ is no longer in use so we re-use it to read/discard.
-    return parseExtensions(io_handle, reinterpret_cast<uint8_t*>(buf_), sizeof(buf_), nullptr);
-  }
-
-  // Initialize the buf_tlv_ only when we need to read the TLVs.
-  if (buf_tlv_.empty()) {
-    buf_tlv_.resize(proxy_protocol_header_.value().extensions_length_);
+ReadOrParseState Filter::readExtensions(Network::ListenerFilterBuffer& buffer) {
+  auto raw_slice = buffer.rawSlice();
+  // waiting for more data if there is no enough data for extensions.
+  if (raw_slice.len_ < (proxy_protocol_header_.value().wholeHeaderLength())) {
+    return ReadOrParseState::TryAgainLater;
   }
 
-  // Parse until we have all the TLVs in buf_tlv.
-  const ReadOrParseState parse_extensions_state =
-      parseExtensions(io_handle, buf_tlv_.data(), buf_tlv_.size(), &buf_tlv_off_);
-  if (parse_extensions_state != ReadOrParseState::Done) {
-    return parse_extensions_state;
+  if (proxy_protocol_header_.value().local_command_ || 0 == config_->numberOfNeededTlvTypes()) {
+    // Ignores the extensions if this is a local command or there's no TLV needs to be saved
+    // to metadata. Those will drained from the buffer in the end.
+    return ReadOrParseState::Done;
   }
 
-  if (!parseTlvs(buf_tlv_)) {
+  const uint8_t* buf = static_cast<const uint8_t*>(raw_slice.mem_) +
```

Comment:
Could a malicous proxy header cause this to do an out-of-bounds read? Should there be a check on the length of the buffer here?

Reply:
This is checked at https://github.com/envoyproxy/envoy/pull/17395/files/ec80e8825245aa0ed651ec63c587aa77e3d53c40#diff-a35fc10b1ce97239d02ac1791aac05069eaf253d8efc77cefe93749f901e36abR368-R369

But I found another mistake https://github.com/envoyproxy/envoy/pull/17395/files/ec80e8825245aa0ed651ec63c587aa77e3d53c40#diff-84c1ba56d718c40a4f8cb0204c28b8a627a9d644775c80af72a37b5d62ef70f2R90
This maxReadBytes should include the max length of extensions. I will correct hat.